src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- hasRightVertex x ('rightVertex' x) == True
+-- hasRightVertex x ('leftVertex' x)  == False
+-- hasRightVertex 1 ('rightVertex' 2) == False
+-- @
+hasRightVertex :: Ord b => b -> AdjacencyMap a b -> Bool
+hasRightVertex y (BAM _ rl) = y `Map.member` rl
+
+-- | Check if a graph contains a given vertex in the given part.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasVertex x 'empty'                   == False
+-- hasVertex (Right x) ('rightVertex' x) == True
+-- hasVertex (Right x) ('leftVertex' x)  == False
+-- hasVertex (Left 1) ('leftVertex' 2)   == False
+-- hasVertex . Left                    == 'hasLeftVertex'
+-- hasVertex . Right                   == 'hasRightVertex'
+-- @
+hasVertex :: (Ord a, Ord b) => Either a b -> AdjacencyMap a b -> Bool
+hasVertex (Left x)  = hasLeftVertex x
+hasVertex (Right y) = hasRightVertex y
+
+-- | Check if a graph is empty.
+-- Complecity: /O(1)/ time.
+--
+-- @
+-- isEmpty 'empty'                 == True
+-- isEmpty ('overlay' 'empty' 'empty') == True
+-- isEmpty ('vertex' x)            == False
+-- isEmpty                       == (==) 'empty'
+-- @
+isEmpty :: AdjacencyMap a b -> Bool
+isEmpty (BAM lr rl) = Map.null lr && Map.null rl
+
+-- | The number of vertices in the left part in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- leftVertexCount 'empty'         == 0
+-- leftVertexCount 'leftVertex' x  == 1
+-- leftVertexCount 'rightVertex' x == 0
+-- leftVertexCount . 'edges'       == 'length' . 'Data.List.nub' . 'map' 'fst'
+-- @
+leftVertexCount :: AdjacencyMap a b -> Int
+leftVertexCount = Map.size . leftAdjacencyMap
+
+-- | The number of vertices in the right part in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- rightVertexCount 'empty'         == 0
+-- rightVertexCount 'rightVertex' x == 1
+-- rightVertexCount 'leftVertex' x  == 0
+-- rightVertexCount . 'edges'       == 'length' . 'Data.List.nub' . 'map' 'snd'
+-- @
+rightVertexCount :: AdjacencyMap a b -> Int
+rightVertexCount = Map.size . rightAdjacencyMap
+
+-- | The number of vertices in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexCount 'empty'         == 0
+-- vertexCount 'leftVertex' x  == 1
+-- vertexCount 'rightVertex' x == 1
+-- vertexCount g             == 'leftVertexCount' g + 'rightVertexCount' g
+-- @
+vertexCount :: AdjacencyMap a b -> Int
+vertexCount g = leftVertexCount g + rightVertexCount g
+
+-- | The number of edges in a graph.
+-- Complexity: /O(n)/ time.
+--
+-- @
+-- edgeCount 'empty'      == 0
+-- edgeCount ('vertex' x) == 0
+-- edgeCount ('edge' x y) == 1
+-- edgeCount . 'edges'    == 'length' . 'Data.List.nub'
+-- @
+edgeCount :: AdjacencyMap a b -> Int
+edgeCount = Map.foldr (+) 0 . fmap Set.size . leftAdjacencyMap
+
+-- Adds all edges needed to make the graph undirected.
+-- Complexity: /O(m log n)/.
+--
+-- /Note: this function is for internal use only./
+addReverseEdges :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AM.AdjacencyMap (Either a b)
```

Comment:
I think this function is called `symmetricClosure` in other modules.

Reply:
Perhaps, you could even use `symmetricClosure` directly?

Reply:
Yup, I just didn't notice it when I tried to search for something like this.